src/v/utils/process.h

```diff
@@ -0,0 +1,148 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "gate_guard.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/with_timeout.hh>
+#include <seastar/util/process.hh>
+
+#include <chrono>
+#include <filesystem>
+#include <system_error>
+#include <utility>
+#include <vector>
+
+namespace ssx {
+
+enum class process_errc : int16_t {
+    success = 0, // must be 0
+    does_not_exist,
+    running,
+    non_zero_exit,
+    signaled
+};
+
+struct process_errc_category final : public std::error_category {
+    const char* name() const noexcept final { return "process::errc"; }
+
+    std::string message(int c) const final {
+        switch (static_cast<process_errc>(c)) {
+        case process_errc::success:
+            return "Success";
+        case process_errc::does_not_exist:
+            return "Process does not exist";
+        case process_errc::running:
+            return "Process is already running";
+        case process_errc::non_zero_exit:
+            return "Process exited with non-zero status";
+        case process_errc::signaled:
+            return "Process exited on signal";
+        }
+        return "process::process_errc::unknown";
+    }
+};
+inline const std::error_category& error_category() noexcept {
+    static process_errc_category e;
+    return e;
+}
+inline std::error_code make_error_code(process_errc e) noexcept {
+    return std::error_code(static_cast<int>(e), error_category());
+}
+
+inline process_errc to_process_errc(std::error_code ec) {
+    return process_errc(ec.value());
+}
+
+class process {
+public:
+    process() {}
+    ~process();
+
+    // Spawns a posix fork and starts running the command.
+    //
+    // Output from stdout and stderr are logged at TRACE level.
+    ss::future<std::error_code> spawn(
```

Comment:
Should this be static and return a result of process or std::error_code?

It cleans up the API because now you can't do invalid things like try to spawn multiple things or have an uninitialized process.

FWIW you see this in other languages for example in Java's ProcessBuilder and Process APIs: https://docs.oracle.com/javase/8/docs/api/java/lang/ProcessBuilder.html

Reply:
notes from sync: can be addressed after adding the factory method.

Reply:
Here's more rationale for this: https://abseil.io/tips/42